cloudinit/net/ephemeral.py

```diff
@@ -152,127 +140,71 @@ def _bringup_device(self):
+                partial(
+                    self.distro.net_ops.link_down,
                     self.interface,
-                    "down",
-                ]
+                    family="inet",
+                )
             )
             self.cleanup_cmds.append(
-                [
-                    "ip",
-                    "-family",
-                    "inet",
-                    "addr",
-                    "del",
-                    cidr,
-                    "dev",
-                    self.interface,
-                ]
+                partial(self.distro.net_ops.del_addr, self.interface, cidr)
             )
 
     def _bringup_static_routes(self):
         # static_routes = [("169.254.169.254/32", "130.56.248.255"),
         #                  ("0.0.0.0/0", "130.56.240.1")]
         for net_address, gateway in self.static_routes:
-            via_arg = []
-            if gateway != "0.0.0.0":
-                via_arg = ["via", gateway]
-
             # Use "append" rather than "add" since the DHCP server may provide
             # rfc3442 classless static routes with multiple routes to the same
             # subnet via different routers or local interface addresses.
             #
             # In this scenario, `ip r add` fails.
             #
             # RHBZ: #2003231
-            subp.subp(
-                ["ip", "-4", "route", "append", net_address]
-                + via_arg
-                + ["dev", self.interface],
-                capture=True,
+            self.distro.net_ops.append_route(
+                self.interface, net_address, gateway
             )
             self.cleanup_cmds.insert(
                 0,
-                ["ip", "-4", "route", "del", net_address]
-                + via_arg
-                + ["dev", self.interface],
+                partial(
+                    self.distro.net_ops.del_route,
+                    self.interface,
+                    net_address,
+                    gateway=gateway,
+                ),
             )
 
     def _bringup_router(self):
         """Perform the ip commands to fully setup the router if needed."""
         # Check if a default route exists and exit if it does
-        out, _ = subp.subp(["ip", "route", "show", "0.0.0.0/0"], capture=True)
+        out = self.distro.net_ops.get_default_route()
         if "default" in out:
             LOG.debug(
                 "Skip ephemeral route setup. %s already has default route: %s",
                 self.interface,
                 out.strip(),
             )
             return
-        subp.subp(
-            [
-                "ip",
-                "-4",
-                "route",
-                "add",
-                self.router,
-                "dev",
-                self.interface,
-                "src",
-                self.ip,
-            ],
-            capture=True,
+        self.distro.net_ops.add_route(
+            self.interface, self.router, source_address=self.ip
```

Comment:
what exactly does this do?
how is this route different from one without `src self.ip`?

Reply:
I think ading `src self.ip` allows the route to force outgoing traffic to use the specified source address, which is otherwise unopinionated by a route that only specifies an interface.